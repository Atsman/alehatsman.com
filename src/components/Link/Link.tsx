import React, { FC } from 'react'
import styled from '@emotion/styled'
import { SerializedStyles } from '@emotion/serialize'
import { Link as GatsbyLink } from 'gatsby'
import {
  typography,
  space,
  layout,
  flexbox,
  LayoutProps,
  TypographyProps,
  SpaceProps
} from 'styled-system'
import { OutboundLink } from 'gatsby-plugin-google-analytics'

const isAnchor = (href: string) => {
  return /^#.+/.test(href)
}

const isExternal = (href: string) => {
  return /^https?:\/\//i.test(href) || /^mailto:.+/.test(href)
}

interface Props extends LayoutProps, TypographyProps, SpaceProps {
  href: string
  className?: string
  css?: SerializedStyles
  target?: string
}

export const Link: FC<Props> = styled(({ children, href, ...props }) => {
  return isExternal(href) || isAnchor(href)
    ? (
    <OutboundLink
      href={href}
      target={isAnchor(href) ? '_self' : '_blank'}
      {...props}
    >
      {children}
    </OutboundLink>
      )
    : (
    <GatsbyLink to={href} {...props}>
      {children}
    </GatsbyLink>
      )
})(typography, space, layout, flexbox)
