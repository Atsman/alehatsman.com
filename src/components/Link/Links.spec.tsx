import React from 'react'
import { render } from '@testing-library/react'
import { Link } from './Link'

describe('Link', () => {
  describe('should render OutboundLink for external links', () => {
    it('should match snapshot', () => {
      const fragment = render(<Link href="https://google.com">Google</Link>).asFragment()
      expect(fragment).toMatchSnapshot()
    })

    it('external link', async () => {
      const { findByText } = render(<Link href="https://google.com">google</Link>)
      const link = await findByText('google')
      expect(link).toBeInTheDocument()
      expect(link).toHaveAttribute('href', 'https://google.com')
      expect(link).toHaveAttribute('target', '_blank')
    })

    it('email', async () => {
      const { findByText } = render(<Link href="mailto:google@google.com">email</Link>)
      const link = await findByText('email')
      expect(link).toBeInTheDocument()
      expect(link).toHaveAttribute('href', 'mailto:google@google.com')
      expect(link).toHaveAttribute('target', '_blank')
    })

    it('anchor', async () => {
      const { findByText } = render(<Link href="#section">google</Link>)
      const link = await findByText('google')
      expect(link).toBeInTheDocument()
      expect(link).toHaveAttribute('href', '#section')
      expect(link).toHaveAttribute('target', '_self')
    })
  })

  describe('should render GatsbyLink for relative links', () => {
    it('should match snapshot', () => {
      const fragment = render(<Link href="/about.html">About</Link>).asFragment()
      expect(fragment).toMatchSnapshot()
    })

    const testCases = [
      '/post/article.html'
    ]
    testCases.forEach(href => {
      it(`${href}`, async () => {
        const { findByText } = render(<Link href={href}>internal</Link>)
        const link = await findByText('internal')
        expect(link).toHaveAttribute('href', href)
      })
    })
  })
})
