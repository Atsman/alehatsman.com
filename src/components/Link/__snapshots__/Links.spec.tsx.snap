// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render GatsbyLink for relative links should match snapshot 1`] = `
<DocumentFragment>
  <a
    class="emotion-0 emotion-1"
    href="/about.html"
  >
    About
  </a>
</DocumentFragment>
`;

exports[`Link should render OutboundLink for external links should match snapshot 1`] = `
<DocumentFragment>
  <a
    class="emotion-0 emotion-1"
    href="https://google.com"
    target="_blank"
  >
    Google
  </a>
</DocumentFragment>
`;
